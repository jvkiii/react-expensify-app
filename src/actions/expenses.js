
import uuid from 'uuid'; //generate universal random id

export const addExpense = (obj = {}) => ({
   type: 'ADD_EXPENSE', 
   expense: {
		id:uuid(),
		description: obj.description || '',
		note: obj.note || '',
		amount: obj.amount || 0,
		createdAt: obj.createdAt || 0
   } 
})

export const editExpense = (id, obj) => ({
	type: 'EDIT_EXPENSE', 
	id,
   updates: obj
})

export const removeExpense = (id = 0) => ({
   type: 'REMOVE_EXPENSE', 
   id
})