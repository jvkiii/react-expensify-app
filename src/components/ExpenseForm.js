import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';


export default class ExpenseForm extends React.Component {
   constructor(props) {
      super(props);

      this.state = {
         description: props.expense ? props.expense.description : '',
         note: props.expense ? props.expense.note : '',
         amount: props.expense ? String(props.expense.amount / 100) : '',
         createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
         calendarFocused: false,
         error:''
      }
   }

   onDescriptionChange = (e) => {
      const description = e.target.value;

      this.setState(() => ({
         description:description
      }))
   }

   onNoteChange = (e) => {
      const note = e.target.value;

      this.setState(() => ({
         note:note
      }))
   }

   onAmountChange = (e) => {
      const amount = e.target.value;

      //validate format: 1234.12
      if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
         this.setState(() => ({
            amount:amount
         }))
      }
   }

   onFormSubmit = e => {
      e.preventDefault();

      if (!this.state.description || !this.state.amount) {
         this.setState(() => ({
            error: 'Please enter a description and amount'
         }))
      } else {
         this.setState(() => ({
            error: ''
         }))
      }

      //send action
      this.props.onFormSubmit({
         description:this.state.description,
         amount:Number(this.state.amount) * 100, //in cents
         note:this.state.note,
         createdAt:this.state.createdAt.valueOf()
      })

      //clear fields
      this.setState(() => ({
         description: '',
         amount:'',
         note:'',
         createdAt: moment()
      }))
   }

   //-- date picker handlers --/

   onDateChange = createdAt => {
      if (createdAt) {
         this.setState(() => ({ createdAt }))
      }
   }

   onCalendarFocusChange = ({ focused }) => {
      this.setState(() => ({ calendarFocused: focused }))
   }

   //-- lifecycle methods --//

   render() {
      return (
         <div>
            {this.state.error && <p>{this.state.error}</p>}
            <form onSubmit={this.onFormSubmit}>
               <input 
                  type="text"
                  placeholder="Description"
                  autoFocus
                  value={this.state.description}
                  onChange={this.onDescriptionChange}
               />
               <input 
                  type="text"
                  placeholder="Amount"
                  value={this.state.amount}
                  onChange={this.onAmountChange}
               />
               <SingleDatePicker
                  id="expense-date-picker" // PropTypes.string.isRequired,
                  date={this.state.createdAt} // momentPropTypes.momentObj or null
                  onDateChange={this.onDateChange} // PropTypes.func.isRequired
                  focused={this.state.calendarFocused} // PropTypes.bool
                  onFocusChange={this.onCalendarFocusChange} // PropTypes.func.isRequired
                  numberOfMonths={1}
                  isOutsideRange={(day) => false}
               />
               <textarea 
                  name=""
                  placeholder="Add a note for your expense (optional)"
                  value={this.state.note}
                  onChange={this.onNoteChange}
               ></textarea>
               <button>Add Expense</button>
            </form>
         </div>
      )
   }
}