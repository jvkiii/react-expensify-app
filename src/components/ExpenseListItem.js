import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ description, amount, createdAt, id}) => (
   <li>
      <Link to={`/edit/${id}`}>
         <strong>{description}</strong>
      </Link>
      <p>Amount: {amount}, Created: {createdAt}</p>
   </li>
);

export default ExpenseListItem;