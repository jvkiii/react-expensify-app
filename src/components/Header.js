import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
   <header>
      <h1>Expensify</h1>
      <NavLink to="/" activeClassName="active" exact={true}>Home</NavLink>
      <NavLink to="/create" activeClassName="active">Add Expense</NavLink>
      <NavLink to="/help" activeClassName="active">Help</NavLink>
   </header>
)

export default Header;