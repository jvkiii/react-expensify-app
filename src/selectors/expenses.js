import moment from 'moment';

export const getFilteredExpenses = (expenses, filters) => { 
	const { text, sortBy, startDate, endDate } = filters;

	const filteredArray = expenses.filter( expense => {
		const createdAtMoment = moment(expense.createdAt);
		const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
		const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
		const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

		return startDateMatch && endDateMatch && textMatch;
	})
	
	const sortedArray = filteredArray.sort((a, b) => {
		if (sortBy === 'date') {
			return a.createdAt < b.createdAt ? 1 : -1;
		}
		else if (sortBy === 'amount') {
			return a.amount < b.amount ? 1 : -1;
		}
	})

	return sortedArray;
}