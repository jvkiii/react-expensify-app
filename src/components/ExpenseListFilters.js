import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

class ExpenseListFilters extends React.Component {
   constructor(props) {
      super(props);

      this.state = {
         calendarFocused:null
      }
   }

   onInputChange = e => {
      this.props.dispatch(setTextFilter(e.target.value));
   }

   onSortChange = e => {
      const value = e.target.value;

      if (value === 'date') {
         this.props.dispatch(sortByDate());
      }
      else if (value === 'amount') {
         this.props.dispatch(sortByAmount());
      }
   }
   
   onDatesChange = ({ startDate, endDate }) => {
      this.props.dispatch(setStartDate(startDate));
      this.props.dispatch(setEndDate(endDate));
   }

   onCalendarFocusChange = calendarFocused => {
      this.setState(() => ({ calendarFocused }));
   }

   render() {
      return (
         <div>
            <input 
               type="text" 
               value={this.props.filters.text}
               onChange={this.onInputChange}
            />
            <select 
               name="sort-by" 
               value={this.props.filters.sortBy}
               onChange={this.onSortChange}
            >
               <option value="date">Date</option>
               <option value="amount">Amount</option>
            </select>

            <DateRangePicker
               startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
               startDateId="start-date-picker" // PropTypes.string.isRequired,
               endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
               endDateId="end-date-picker" // PropTypes.string.isRequired,
               onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
               focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
               onFocusChange={this.onCalendarFocusChange} // PropTypes.func.isRequired,
               numberOfMonths={1}
               isOutsideRange={() => false}
               showClearDates={true}
            />
         </div>
      )
   }
}

const mapStateToProps = state => {
   return {
      filters: state.filters
   }
}

export default connect(mapStateToProps)(ExpenseListFilters);