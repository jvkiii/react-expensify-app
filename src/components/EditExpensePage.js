import React from 'react';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

const EditExpensePage = (props) => (
   <div>
      
      <ExpenseForm 
         expense={props.expense}
         onFormSubmit={ expense => {
            console.log(expense);

            //send action
            props.dispatch(editExpense(props.expense.id, expense));

            //go to homepage
            props.history.push('/');
         }}
      />

      <button onClick={e => {
         props.dispatch(removeExpense(props.expense.id));
         props.history.push('/');
      }}>Remove</button>
   </div>
)

const mapStateToProps = (state, props) => {
   return {
      expense: state.expenses.find(expense => expense.id === props.match.params.id)
   }
};

export default connect(mapStateToProps)(EditExpensePage);