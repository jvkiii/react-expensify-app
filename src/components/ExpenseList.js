
import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import { getFilteredExpenses } from '../selectors/expenses';

export const ExpenseList = (props) => (
   <div>
      { 
         props.expenses.length === 0 ? (
            <p>No expenses</p>
         ) : (
            <ul>
               {props.expenses.map( (expense, i) => {
                  return <ExpenseListItem key={i} {...expense} />
               })}
            </ul>
         )
      }
   </div>
)

const mapStateToProps = state => {
   return {
      expenses: getFilteredExpenses(state.expenses, state.filters)
   }
};

export default connect(mapStateToProps)(ExpenseList);