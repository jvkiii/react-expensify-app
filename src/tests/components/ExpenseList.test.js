import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses'; //sample expenses


test('should render ExpenseList with expenses', () => {
   const wrapper = shallow(<ExpenseList expenses={expenses}/>);
   expect(wrapper).toMatchSnapshot();
})

test('should render ExpenseList with no expenses', () => {
   const wrapper = shallow(<ExpenseList expenses={[]}/>);
   expect(wrapper).toMatchSnapshot();
})