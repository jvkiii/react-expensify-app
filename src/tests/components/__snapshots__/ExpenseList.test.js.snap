// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <ul>
    <ExpenseListItem
      amount={195}
      createdAt={0}
      description="Gum"
      id="1"
      key="0"
      note=""
    />
    <ExpenseListItem
      amount={109500}
      createdAt={-345600000}
      description="Rent"
      id="2"
      key="1"
      note=""
    />
    <ExpenseListItem
      amount={4500}
      createdAt={345600000}
      description="Credit Card"
      id="3"
      key="2"
      note=""
    />
  </ul>
</div>
`;

exports[`should render ExpenseList with no expenses 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;
